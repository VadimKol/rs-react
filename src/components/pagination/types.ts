import { type MouseEventHandler } from 'react';

export interface PaginationProps {
  page: number;
  total: number;
  setPage: (page: number) => void;
  handleClose: MouseEventHandler;
}
