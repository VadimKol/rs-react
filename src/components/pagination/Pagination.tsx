import { type ReactNode } from 'react';

import styles from './styles.module.scss';
import { type PaginationProps } from './types';

export function Pagination({ page, total, setPage, setLoader }: PaginationProps): ReactNode {
  let paginationStyles = styles.pagination_box;
  if (!(page - 1)) {
    paginationStyles = `${styles.pagination_box_before}`;
  }
  if (!(page < total)) {
    paginationStyles = `${styles.pagination_box_after}`;
  }
  if (!(page - 1) && !(page < total)) {
    paginationStyles = styles.pagination_box;
  }
  return (
    <div className={styles.pagination}>
      <div className={paginationStyles}>
        {Boolean(page - 1) && (
          <button
            type="button"
            aria-label="Left"
            className={styles.pag_left}
            onClick={() => {
              setPage(page - 1);
              setLoader(true);
            }}
          />
        )}
        <div className={styles.pages}>
          <span>{page}</span>-<span>{total}</span>
        </div>
        {page < total && (
          <button
            type="button"
            aria-label="Right"
            className={styles.pag_right}
            onClick={() => {
              setPage(page + 1);
              setLoader(true);
            }}
          />
        )}
      </div>
    </div>
  );
}
