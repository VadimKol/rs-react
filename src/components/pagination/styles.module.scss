.pagination {
  display: flex;
  flex-grow: 1;
  justify-content: center;
  align-items: flex-start;
  margin: 30px 0 8px;
}

.pagination_box,
.pagination_box_before,
.pagination_box_after {
  display: flex;
  gap: 15px;
  align-items: center;
}

.pagination_box_before {
  &::before {
    content: "";
    display: block;
    width: 45px;
    height: 45px;
  }
}

.pagination_box_after {
  &::after {
    content: "";
    display: block;
    width: 45px;
    height: 45px;
  }
}

.pages {
  padding: 0 5px;
  font-size: 37.97px;
  color: var(--text-color);
  background-color: var(--background-color);
  border-radius: 6px;
  box-shadow: 0 0 4px 4px var(--background-color);
}

.pag_left {
  transform: rotate(90deg);
}

.pag_right {
  transform: rotate(-90deg);
}

.pag_left,
.pag_right {
  width: 45px;
  height: 45px;
  padding: 0;
  background-color: var(--background-color);
  border: none;
  border-radius: 6px;
  box-shadow: 0 0 4px 4px var(--background-color);
  cursor: pointer;
  transition: transform 0.5s;

  &:focus-visible {
    border-radius: 6px;
    outline: 2px solid var(--primary-color);
  }

  &::after {
    content: "";
    display: block;
    width: 100%;
    height: 100%;
    background-color: var(--text-color);
    transition: background-color 0.5s;
    mask-image: url("/src/assets/images/arrow_down.svg");
    mask-size: cover;
  }

  @media (hover: hover) {
    &:hover::after {
      background-color: var(--primary-color);
    }
  }
}
