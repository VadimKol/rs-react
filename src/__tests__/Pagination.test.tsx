import '@testing-library/jest-dom';

import { fireEvent, render, screen } from '@testing-library/react';
import { useState } from 'react';
import { MemoryRouter } from 'react-router-dom';

import { Pagination } from '@/components/pagination/Pagination';

let mockSearchParam = new URLSearchParams('');

jest.mock('react-router-dom', () => ({
  ...jest.requireActual<object>('react-router-dom'),
  useSearchParams: (): [URLSearchParams, (newParams: URLSearchParams) => void] => {
    const [params, setParams] = useState(mockSearchParam);
    return [
      params,
      (newParams: URLSearchParams): void => {
        mockSearchParam = newParams;
        setParams(newParams);
      },
    ];
  },
}));

describe('Pagination Component', () => {
  it('renders correctly', () => {
    const { container } = render(
      <MemoryRouter>
        <Pagination page={1} total={2} setPage={jest.fn} setLoader={jest.fn} handleClose={jest.fn} />
      </MemoryRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should change page query params', () => {
    render(
      <MemoryRouter>
        <Pagination page={2} total={3} setPage={jest.fn} setLoader={jest.fn} handleClose={jest.fn} />
      </MemoryRouter>,
    );

    const buttonRight = screen.getByTestId('right');
    const buttonLeft = screen.getByTestId('left');

    fireEvent.click(buttonRight);
    expect(mockSearchParam).toEqual({ page: '3' });
    fireEvent.click(buttonLeft);
    expect(mockSearchParam).toEqual({ page: '1' });
  });
});
