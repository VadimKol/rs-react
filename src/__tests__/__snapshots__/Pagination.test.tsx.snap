// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Component renders correctly 1`] = `
<div>
  <div
    class="pagination"
  >
    <div
      class="pagination_box_before"
    >
      <div
        class="pages"
      >
        <span>
          1
        </span>
        -
        <span>
          2
        </span>
      </div>
      <button
        aria-label="Right"
        class="pag_right"
        data-testid="right"
        type="button"
      />
    </div>
  </div>
</div>
`;
